### PR TITLE
Fix factory loading

### PR DESCRIPTION
**Pull request by elia**
**Description**

The current FactoryBot support code doesn't take into account that Solidus factories need to be loaded before app-level factories, to allow modifications.

With this change using `factory_bot_rails` it will work out of the box (the initializer works with v4.8 up to v6).
Nothing is done if `FactoryBot` is used directly, assuming the a more fine grained control/configuration is being used.
The code will be activated only if FactoryBotRails is defined, so to avoid running it in production is enough to keep the gem in the developmen/test groups as advised by the factory_bot readme.

Ref https://github.com/solidusio/solidus/issues/3906 (this issue will be fixed by a back port of this PR to v2.11).

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [ ] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: core/lib/spree/core/engine.rb
Comment (kennyadsl) [inline core/lib/spree/core/engine.rb:98]: I don't understand why we can't use the default way of setting factories for an Engine as described [here](https://github.com/thoughtbot/factory_bot_rails#automatic-factory-definition-loading)? Is this approach excluded for a reason?

**Review by kennyadsl — COMMENTED** [review]
Left a preliminary question about the approach used, thanks @elia!

File: core/lib/spree/core/engine.rb
Comment (elia) [reply core/lib/spree/core/engine.rb:98]: If you're referring to the use of the Rails-style `config.factory_bot.definition_file_paths` that's not available for FactoryBot v4 (don't remember it it was introduced in 5 or 6), by updating directly `FactoryBot.definition_file_paths` we're compatible with all versions in the 4..6 range.

The `defined?` check is necessary because FactoryBot might not be available in some environments (e.g. production) and the check is done on `FactoryBotRails` with the assumption that if you're not using it, you probably also have some custom setup we don't want to mess with. In this way we can rely on the initializer as hook to inject the Solidus paths.

About using an initializer with an `after:` condition, the main reason for this is that custom factories may depend on the solidus factories, so we need a way to enforce loading them before the custom ones. Specifically `FactoryBotRails` has an initializer named `"factory_bot.set_factory_paths"` that will reset the paths, so if we add the solidus paths before that they'll be wiped out by this initializer.

Comment (kennyadsl) [reply core/lib/spree/core/engine.rb:97]: Ok, it makes sense now, thanks! So, next question: why still draft?

Comment (elia) [reply core/lib/spree/core/engine.rb:97]: I was waiting for the CI to be green, but when I pushed last week master was temporarily red, so time passed by 😉 
I just fixed the linting suggestion and repushed, will mark it as ready in a moment 👍

File: core/lib/spree/testing_support/factory_bot.rb
Comment (kennyadsl) [inline core/lib/spree/testing_support/factory_bot.rb:26]: `caller(1)` is giving me more accurate details when running specs in a real store: 

```
# caller(2) example result
DEPRECATION WARNING: Please do not cherry-pick factories, this is not well supported by FactoryBot. Use `require "spree/testing_support/factories"` instead. (called from block in <top (required)> at /Users/kennyadsl/Code/nebulab/solidus/core/lib/spree/testing_support/factories.rb:11)

# caller(1) example result
DEPRECATION WARNING: Please do not cherry-pick factories, this is not well supported by FactoryBot. Use `require "spree/testing_support/factories"` instead. (called from <top (required)> at /Users/kennyadsl/Code/nebulab/solidus/core/lib/spree/testing_support/factories/zone_factory.rb:4)

```

Comment (elia) [reply core/lib/spree/testing_support/factory_bot.rb]: I updated the code to mention the changelog instruction as a way to fix the deprecation and to point it to the first non-core line from the backtrace so ideally it will point right in the place where the cherry-picking is happening (i.e. inside the rails application).

**Comment by tvdeyen** [comment]
FYI: This change and the PR that introduced it breaks every gem/app in our eco system in a minor version that loads individual factories. We probably should revert it in 2.11 or add a fix that still supports loading individual factories and only remove it in 3.0

**Comment by kennyadsl** [comment]
@tvdeyen thanks for the feedback, the goal of this PR is to still provide support for loading individual factories without breaking any application. Do you have any evidence that it is breaking things? I tried this PR with a couple of real stores and they just give deprecation warnings. Of course, I don't expect to have covered all possible scenarios for factory_bot configuration.

**Comment by tvdeyen** [comment]
@kennyadsl I have not tested this PR, but this CI run is from a recent Solidus v2.11

https://github.com/AlchemyCMS/alchemy-solidus/runs/1888940075?check_suite_focus=true

**Comment by tvdeyen** [comment]
@elia @kennyadsl the switch to mutate FactoryBot factory load paths to load factories breaks the current way of requiring individual factories in tests suites of gems and apps. The errors you are getting are missing factories, duplicated factories and missing sequences. I think there is no easy way of supporting both ways the same time.

I agree that the FactoryBot way is better and we should not require individual factories in test suites, but this is an widely used approach and we should figure out a way to a) support both or b) revert this change in 2.11

**Comment by kennyadsl** [comment]
For gems, I made [this PR on Solidus Support](Fix loading factories in extensions after the last changes in Solidus solidusio/solidus_dev_support#169) that, once this is merged, "should" fix all extensions. I only tried with a couple so I'm not really sure it works for everyone. 

For stores, reverting breaking changes is exactly what this PR has been made for. If you can help us understand the exact problem you are facing, we'd be happy to explore it or if not solvable, I agree we should revert and add this breaking behavior in a major.

**Comment by kennyadsl** [comment]
@tvdeyen I'm taking a look at `alchemy-solidus` (checking that out to the commit before your fix in master) with this branch and I think you are right. 

There's one big thing we didn't consider: when cherry-picking a file we are showing a deprecation message but not loading the specific dependency of that factory anymore. 

For example when requiring the [product_factory](https://github.com/solidusio/solidus/blob/70979e8dd1df543f515314a317787f90b28eae1a/core/lib/spree/testing_support/factories/product_factory.rb) individually, we print the deprecation message but we are not re-requiring all the factories that this file was relying on, which were:

```
require 'spree/testing_support/sequences'
require 'spree/testing_support/factories/shipping_category_factory'
require 'spree/testing_support/factories/stock_location_factory'
require 'spree/testing_support/factories/tax_category_factory'
require 'spree/testing_support/factories/product_option_type_factory'
```

I'll talk with @elia to see if there's a way to make it work but it's becoming uncomfortable to keep adding code again and again to fix all the possible cases, so I'd strongly consider a revert. 

Thanks again for pointing this out.

**Comment by kennyadsl** [comment]
@tvdeyen the last pushed code restores backward compatibility to require factories individually, I tested it with alchemy-solidus (checking out to a commit previous to your fix in https://github.com/AlchemyCMS/alchemy-solidus/pull/72). 

I think that after some git commit improvements we can merge and backport this into 2.11. Projects that are requiring factories individually will keep their stores/gems working and we also print a deprecation warning since this behavior will not be supported anymore with Solidus 3.0.

What do you think?

**Comment by tvdeyen** [comment]
@kennyadsl wow. Thanks. This looks promising 🙏🏻 @mamhoff and I will test this and provide feedback asap.

**Review by tvdeyen — APPROVED** [review]
Looks great.

**Review by kennyadsl — APPROVED** [review]
Worked together with @elia here. 

We tested this change on a real application, 2-3 extensions and an external gem (`alchemy-solidus`) that was cherry-picking individual factories. Apparently fixed the backward-compatibility issues in all of them, printing a deprecation warning as well.

**Pull request merged**
